### ISSUE TITLE
[Spike] Building Cloud images with cos-toolkit

### ISSUE THREAD

**Issue by mudler** [rancher-sandbox/cOS-toolkit#174]
**Is your feature request related to a problem? Please describe.**
A way to build cloud images with cos-toolkit

**Describe the solution you'd like**
A helper/workflow to build cloud images (even ISOs) with [KIWI](https://github.com/OSInside/kiwi). It has to be:
- CI friendly ( need to attach to GH, but also to Drone pipelines, for example )
- preferably with a helper script, or even better embedded in the #108 CLI

**Describe alternatives you've considered**

**Additional context**

**Comment by Itxaka** [comment]
Makes more sense to use packer here as we are already building virtualbox and qemu images with it? In the end, the way packer create the image is the "official" one (run a vm, snapshot it, there is your image)

**Comment by davidcassany** [comment]
Here is some basic settings of the EC2, GCE and Azure images for SUSE. There are some similarities as well as some differences in each:

## Azure:

- firmware -> uefi
- efipart size -> 512Mb
- boot partition -> 1G
- disk format -> vhd-fixed with certain flags (force_size and vhdfixedtag)
- device persistency is set to UUID (I'd say this is irrelevant as it affects mostly fstab creation)
- kernel params -> `rootdelay=300 net.ifnames=0 earlyprintk=ttyS0 USE_BY_UUID_DEVICE_NAMES=1 dis_ucode_ldr csi_mod.use_blk_mq=1 multipath=off` (we should investigate the meaning of all these params)
- bootloader -> grub2 on serial console
- disk size -> 30G
- expansion strategy: 
- cloud initialization technology: cloud-init
- extra packages for Azure (clear use case for cOS flavours concept):
  - kernel-azure 
  - azure-cli
  - hyper-v
  - lsscsi
  - python3-azuremetadata (I'd say we already have that in YIP)
  - shim
**Instance initialization**
  - cloud-init
  - cloud-init-config-suse (need to investigate if suse configs are compatible with yip)
  - python-azure-agent (need to verify what this provides, is it a hard requirement?)
**Instance registration**
  - cloud-regionsrv-client-plugin-azure
  - regionServiceClientConfigAzure
  - regionServiceCertsAzure


## Google Cloud

* firmware -> uefi
* efipart size -> default
* boot partition -> none
* disk format -> gce (gcelicense attribute)
* device persistency is set to LABEL
* kernel params -> console=ttyS0,38400n8 net.ifnames=0 dis_ucode_ldr multipath=off
* bootloader -> grub2 on serial console
* disk size 10G
* expansion strategy: growpart
* cloud initialization technology: google guest agent!
* extra packages for Google Cloud:
  - cloud-netconfig-gce (is this requried for early boot? or it is devoted to main OS?)
  - growpart
  - growpart-rootgrow
  - kernel-default
  - python3-gcemetadata (I guess this is already provoided in YIP, need to check the boot stage though)
  - shim 
**Instance initialization**
  - google-guest-agent
  - google-guest-configs
  - google-guest-oslogin
  - google-osconfig-agent
**Instance register**
  - cloud-regionsrv-client-plugin-gce
  - regionServiceClientConfigGCE
  - regionServiceCertsGCE
  
## AWS (ec2-hvm):

* firmware -> efi
* efipart size -> default
* boot partition -> none
* disk format -> raw
* device persistency is set to LABEL
* kernel params -> console=ttyS0 net.ifnames=0 nvme_core.io_timeout=4294967295 nvme_core.admin_timeout=4294967295 8250.nr_uarts=4 dis_ucode_ldr multipath=off
* bootloader -> grub2 
* disk size 10G
* xen_loader="hvmloader"
* expansion strategy: 
* cloud initialization technology: cloud-init
* extra packages for Google Cloud:
  - amazon-ssm-agent
  - aws-cli
  - cloud-netconfig-ec2
  - kernel-default
  - python3-ec2imgutils
  - python3-ec2metadata (I guess this is already provoided in YIP, need to check the boot stage though)
  - xen-libs
  - xen-tools-domU
  **Instance initialization**
  - cloud-init
  - cloud-init-config-suse
  **Instance register**
  - cloud-regionsrv-client-plugin-ec2
  - regionServiceClientConfigEC2
  - regionServiceCertsEC2


At a glance the simplest images seams to be the AWS, as there is no secure boot, no boot partition and no custom EFI partition size.

**Comment by davidcassany** [comment]
Regarding the image layout I'd say we should mostly provide a legacy MBR partition (around 2MB) at the beginning, then the EFI DOS partition and then the COS_RECOVERY or COS_OEM partition and finally the COS_STATE. No COS_PERSISTENT at all, I would create that on firstboot.

The reasoning behind is that we should consider the possibility of dynamically increasing the size for COS_STATE at boot time based on some cloud-init config if needed. The user might choose a huge disk for the instance on the cloud provider and I believe having a way to balance this storage between COS_STATE and COS_PERSISTENT partitions is relevant, probably COS_PERSISTENT it could even be a remote volume.

So IMHO the layout could be:

1. MBR legacy partition to allow legacy bios boot
2. EFI partition that points to the appropriate grub.cfg
3. COS_OEM partition
4. COS_RECOVERY
5. COS_STATE
6. unpartitioned space to be assinged at first boot to COS_STATE or COS_PERSISTENT according to some config or defaults.

I personally would include the grub configuration, the main `grub.cfg` into the COS_RECOVERY partition, so we enforce the possibility to actually kill the COS_STATE partition and one can still recover from such a mess.

COS_SYSTEM loop device I would base it on a compressed squasfs image, so it the partition can be pretty small. I would not embrace the immutability concept there, meaning I would just run it as it was a live image (immutable by design).

COS_STATE I would just size it to hold some active.img and passive.img and assume one can setup and instance that expands it.

I would always allow legacy bios boot, so all our images could be used with EFI or bios. It is not hard to do it and results in a no brainer setup.

**Comment by mudler** [comment]
A quick search on packer (just for sake of completeness) brought up:

- Amazon: http://work.haufegroup.io/automate-ami-with-packer/
- GCE: https://www.packer.io/docs/post-processors/googlecompute-import#qemu-builder-example (from QEMU image to gce)
- Azure: https://github.com/MicrosoftDocs/azure-docs/blob/master/articles/virtual-machines/windows/build-image-with-packer.md- a use case at hand: https://github.com/garvincasimir/Azure-ISO-To-VHD

**Comment by mudler** [comment]
I will try to summarize my general impressions from the articles so far:

The packer approach might be faster to implement, but at least, speaking for what I've seen for Azure seems quite a bit hackish. On the other hand I'm tempted to say that is more _portable_ in the sense that if the process runs in QEMU & similar, might be "easier" to run

The kiwi approach might be harder to implement, and might require more changes from our side but it looks more polished - it might also have some drawbacks with caps/privileged.

**Comment by davidcassany** [comment]
>     * Amazon: http://work.haufegroup.io/automate-ami-with-packer/

At a glance this feature is mostly covering https://docs.aws.amazon.com/vm-import/latest/userguide/vmimport-image-import.html which is out of scope I'd say. I doubt we want to include the image upload to AWS workflow, we just care the image binary for now. From what I understand, packer does not provide any AWS specifics at build time, it relies on the installer media only.



>     * GCE: https://www.packer.io/docs/post-processors/googlecompute-import#qemu-builder-example (from QEMU image to gce)

Similar as with AWS, this mostly covering the import procedure. However packer has [Google Compute builder](https://www.packer.io/docs/builders/googlecompute) that seams to _transform_ a given image to GCE (I could not really figure out what transform means). To build GCE from scratch Packer points [here](https://cloud.google.com/compute/docs/import/import-existing-image)



>     * Azure: https://github.com/MicrosoftDocs/azure-docs/blob/master/articles/virtual-machines/windows/build-image-with-packer.md

This seams to be wrapper around [azure-arm](https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/overview) so basically it is also providing the upload/import feature.

So basically these procedures are mostly providing an upload & import procedure to cloud providers, I think this is out of topic for now. 

I understand the idea is building image binaries that can be directly uploaded to the cloud provider without modifications. How to upload them is not our business now.

Up to my understanding Packer has little to do in regards to that, since the image creation is mostly based on the installation media and its applications. So using Packer means translating the relevant details of https://github.com/rancher-sandbox/cOS-toolkit/issues/174#issuecomment-843322156 into `cos-installer` and into the containerized appliance (`system/cos`) depending on the feature/config.

**Comment by mudler** [comment]
> > ```
> > * Amazon: http://work.haufegroup.io/automate-ami-with-packer/
> > ```
> 
> At a glance this feature is mostly covering https://docs.aws.amazon.com/vm-import/latest/userguide/vmimport-image-import.html which is out of scope I'd say. I doubt we want to include the image upload to AWS workflow, we just care the image binary for now. From what I understand, packer does not provide any AWS specifics at build time, it relies on the installer media only.
> 
> > ```
> > * GCE: https://www.packer.io/docs/post-processors/googlecompute-import#qemu-builder-example (from QEMU image to gce)
> > ```
> 
> Similar as with AWS, this mostly covering the import procedure. However packer has [Google Compute builder](https://www.packer.io/docs/builders/googlecompute) that seams to _transform_ a given image to GCE (I could not really figure out what transform means). To build GCE from scratch Packer points [here](https://cloud.google.com/compute/docs/import/import-existing-image)
> 
> > ```
> > * Azure: https://github.com/MicrosoftDocs/azure-docs/blob/master/articles/virtual-machines/windows/build-image-with-packer.md
> > ```
> 
> This seams to be wrapper around [azure-arm](https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/overview) so basically it is also providing the upload/import feature.
> 
> So basically these procedures are mostly providing an upload & import procedure to cloud providers, I think this is out of topic for now.
> 
> I understand the idea is building image binaries that can be directly uploaded to the cloud provider without modifications. How to upload them is not our business now.

Yes, all the solution for scratch images around Packer require a bootable medium, and end up as an import.

How to upload them it's not our business I agree, but a workflow to get the image is, and typically a user end up to want to upload those in the cloud, or use them for e.g. from an AWS account. I don't think Packer is out of scope here, because at the end of the day you still have an image to boot up with different providers, it's just that Packer does that in a _different_ way

**Comment by Itxaka** [comment]
I think one important thing that is not discussed here is the difference between ease of building for developers/testers/users.

If im not mistaken building cloud images with packer requires you to have accounts at the cloud providers as they are all booted on a vm, configured then snapshoted, while kiwi just builds everything locally without access to external services, so that is a big plus for a testing playground or locally building images for testing. I dont want to have credentials for all cloud providers if Im building an iso to test different things nor I want to launch any vms at all!

Plus Im not sure building in vms and snapshotting allows us to have reproducible isos? like md5 exact isos?

**Comment by mudler** [comment]
> I think one important thing that is not discussed here is the difference between ease of building for developers/testers/users.
> 
> If im not mistaken building cloud images with packer requires you to have accounts at the cloud providers as they are all booted on a vm, configured then snapshoted, while kiwi just builds everything locally without access to external services, so that is a big plus for a testing playground or locally building images for testing. I dont want to have credentials for all cloud providers if Im building an iso to test different things nor I want to launch any vms at all!
> 
> Plus Im not sure building in vms and snapshotting allows us to have reproducible isos? like md5 exact isos?

I don't think we should just limit to "one way" in this sense. We need a separate story for the "Vanilla images" that we are going to maintain that can be used to switch later to its own flavor. This will avoid majority of users to go into the deep of building a cloud image DIY. What we need to be sure is to provide a way of doing that, so I would just propose to split the use cases in different story and go over with them on a need basis. 

I see your point, indeed the packer templates are very handy, because depending on the personas, having API keys at hand is not an issue. E.g. a user might be already familiar with packer, and is more complex checking out how we build things that rather running a simple template.

From an installation point of view running templates to get a custom image on a Cloud doesn't need exact md5sums from us. It is compared to a "replicate" install or a golden image which can be later reused, it is mostly an entrypoint for hacking around and get over if you need a simple thing from it.

**Comment by davidcassany** [comment]
> Plus Im not sure building in vms and snapshotting allows us to have reproducible isos? like md5 exact isos?

Having bit a bit reproducible images would be a nice feature, I am not sure if you attended the E&I talk of yesterday about reproducible RPMs (bit a bit). It is nightmare just for a mere RPM, images are just another level or league... :sweat_smile:  Anyway, having a clear view of to what extend our images are reproducible is an important matter. I'd relay on the job done for within the verification to asses images are reproducible.

**Comment by davidcassany** [comment]
Closing since most of the spike is already described in this issue comments and also as part of #217. There is a [wiki page](https://github.com/rancher-sandbox/cOS-toolkit/wiki/Images-for-public-cloud-providers) linking the discussion of this issue. I consider this issue done.
